File: backend/app/routes/transactions.py
from datetime import date, timedelta

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from ..database import get_db
from ..dependencies import require_user_or_admin
from ..models import Book, Membership, Transaction, User
from ..schemas import IssueBookRequest, PayFineRequest, ReturnBookRequest

router = APIRouter(prefix="/transactions", tags=["Transactions"])
FINE_PER_DAY = 10


def _has_unpaid_fine(db: Session, user_id: int) -> bool:
    txn = (
        db.query(Transaction)
        .filter(
            Transaction.user_id == user_id,
            Transaction.calculated_fine > Transaction.fine_paid,
        )
        .first()
    )
    return txn is not None


def _has_active_membership(db: Session, user: User, on_date: date) -> bool:
    if not user.membership_id:
        return False
    membership = db.query(Membership).filter(Membership.id == user.membership_id).first()
    if not membership:
        return False
    return membership.active and membership.start_date <= on_date <= membership.end_date


@router.get("/book-available")
def book_available(
    title: str | None = Query(default=None),
    media_type: str | None = Query(default=None),
    db: Session = Depends(get_db),
    _: User = Depends(require_user_or_admin),
):
    if not title and not media_type:
        raise HTTPException(status_code=400, detail="Provide either title or media_type")

    query = db.query(Book).filter(Book.available == True)
    if title:
        query = query.filter(Book.title.ilike(f"%{title.strip()}%"))
    if media_type:
        query = query.filter(Book.media_type == media_type.strip().lower())

    books = query.all()
    return [
        {
            "id": b.id,
            "title": b.title,
            "author": b.author,
            "serial_no": b.serial_no,
            "media_type": b.media_type,
            "available": b.available,
        }
        for b in books
    ]


@router.post("/issue-book")
def issue_book(
    payload: IssueBookRequest,
    db: Session = Depends(get_db),
    _: User = Depends(require_user_or_admin),
):
    user = db.query(User).filter(User.id == payload.user_id).first()
    if not user:
        raise HTTPException(status_code=404, detail="User not found")

    book = db.query(Book).filter(Book.id == payload.book_id, Book.available == True).first()
    if not book:
        raise HTTPException(status_code=400, detail="Book not available")

    if payload.issue_date < date.today():
        raise HTTPException(status_code=400, detail="Issue date cannot be lesser than today")

    if not _has_active_membership(db, user, payload.issue_date):
        raise HTTPException(status_code=400, detail="Active membership required")

    if _has_unpaid_fine(db, payload.user_id):
        raise HTTPException(status_code=400, detail="User has unpaid fine")

    max_return_date = payload.issue_date + timedelta(days=15)
    due_date = payload.return_date or max_return_date

    if due_date > max_return_date:
        raise HTTPException(status_code=400, detail="Return date cannot be greater than 15 days")
    if due_date < payload.issue_date:
        raise HTTPException(status_code=400, detail="Return date cannot be before issue date")

    txn = Transaction(
        book_id=book.id,
        user_id=payload.user_id,
        issue_date=payload.issue_date,
        due_date=due_date,
        remarks=payload.remarks,
    )
    book.available = False
    db.add(txn)
    db.commit()
    db.refresh(txn)
    return {
        "message": "Book issued successfully",
        "transaction_id": txn.id,
        "book_name": book.title,
        "author": book.author,
        "issue_date": txn.issue_date,
        "return_date": txn.due_date,
    }


@router.post("/return-book")
def return_book(
    payload: ReturnBookRequest,
    db: Session = Depends(get_db),
    _: User = Depends(require_user_or_admin),
):
    txn = (
        db.query(Transaction)
        .filter(Transaction.id == payload.transaction_id, Transaction.return_date == None)
        .first()
    )
    if not txn:
        raise HTTPException(status_code=404, detail="Active transaction not found")

    book = db.query(Book).filter(Book.id == txn.book_id).first()
    if not book:
        raise HTTPException(status_code=404, detail="Book not found")
    if book.serial_no != payload.serial_no:
        raise HTTPException(status_code=400, detail="Serial number mismatch")

    txn.pending_return_date = payload.return_date
    days_late = max((payload.return_date - txn.due_date).days, 0)
    fine = days_late * FINE_PER_DAY
    txn.calculated_fine = fine
    db.commit()
    return {
        "message": "Proceed to pay fine page",
        "transaction_id": txn.id,
        "book_name": book.title,
        "author": book.author,
        "issue_date": txn.issue_date,
        "return_date": txn.due_date,
        "selected_return_date": payload.return_date,
        "fine": fine,
    }


@router.post("/pay-fine")
def pay_fine(
    payload: PayFineRequest,
    db: Session = Depends(get_db),
    _: User = Depends(require_user_or_admin),
):
    txn = (
        db.query(Transaction)
        .filter(Transaction.id == payload.transaction_id, Transaction.return_date == None)
        .first()
    )
    if not txn:
        raise HTTPException(status_code=404, detail="Transaction not found")

    if txn.calculated_fine > 0 and not payload.fine_paid:
        raise HTTPException(status_code=400, detail="Paid fine checkbox is mandatory for pending fine")

    txn.fine_paid = txn.calculated_fine if payload.fine_paid else 0
    txn.return_date = txn.pending_return_date or date.today()
    txn.pending_return_date = None
    if payload.remarks:
        txn.remarks = payload.remarks

    book = db.query(Book).filter(Book.id == txn.book_id).first()
    if book:
        book.available = True

    db.commit()
    return {"message": "Book returned successfully"}


@router.get("/overdue-returns")
def overdue_returns(db: Session = Depends(get_db), _: User = Depends(require_user_or_admin)):
    today = date.today()
    txns = (
        db.query(Transaction)
        .filter(Transaction.return_date == None, Transaction.due_date < today)
        .all()
    )
    return [
        {
            "transaction_id": t.id,
            "book_id": t.book_id,
            "user_id": t.user_id,
            "due_date": t.due_date,
            "days_late": (today - t.due_date).days,
            "fine": (today - t.due_date).days * FINE_PER_DAY,
        }
        for t in txns
    ]


@router.get("/active-issues")
def active_issues(db: Session = Depends(get_db), _: User = Depends(require_user_or_admin)):
    txns = db.query(Transaction).filter(Transaction.return_date == None).all()
    return [
        {
            "transaction_id": t.id,
            "user_id": t.user_id,
            "book_id": t.book_id,
            "issue_date": t.issue_date,
            "due_date": t.due_date,
        }
        for t in txns
    ]

File: frontend/js/transactions.js
const API = "http://127.0.0.1:8000";

function authHeaders() {
  return {
    "Content-Type": "application/json",
    Authorization: "Bearer " + localStorage.getItem("token"),
  };
}

function setDashboardLink() {
  const role = localStorage.getItem("role");
  const link = document.querySelector(".nav-links a");
  if (!link) return;
  link.href = role === "admin" ? "../admin_home.html" : "../user_home.html";
}

function todayISO() {
  return new Date().toISOString().slice(0, 10);
}

function addDaysISO(days) {
  const d = new Date();
  d.setDate(d.getDate() + days);
  return d.toISOString().slice(0, 10);
}

async function searchAvailableBooks() {
  const title = document.getElementById("search_title")?.value?.trim();
  const mediaType = document.querySelector('input[name="media_type"]:checked')?.value;
  const msg = document.getElementById("msg");

  if (!title && !mediaType) {
    msg.innerText = "Enter book name or select type before search.";
    return;
  }

  const query = new URLSearchParams();
  if (title) query.append("title", title);
  if (mediaType) query.append("media_type", mediaType);

  const res = await fetch(`${API}/transactions/book-available?${query.toString()}`, {
    headers: authHeaders(),
  });
  const data = await res.json();
  const table = document.getElementById("booksTable");
  table.innerHTML = `
    <tr>
      <th>Select</th>
      <th>Book</th>
      <th>Author</th>
      <th>Serial No</th>
      <th>Type</th>
    </tr>`;

  if (!res.ok) {
    msg.innerText = data.detail || "Unable to fetch books";
    return;
  }

  if (!data.length) {
    msg.innerText = "No books found.";
    return;
  }

  data.forEach((book) => {
    table.innerHTML += `
      <tr>
        <td><input type="radio" name="book" value="${book.id}" data-title="${book.title}" data-author="${book.author}"></td>
        <td>${book.title}</td>
        <td>${book.author}</td>
        <td>${book.serial_no}</td>
        <td>${book.media_type}</td>
      </tr>`;
  });

  document.querySelectorAll('input[name="book"]').forEach((r) => {
    r.addEventListener("change", (e) => {
      document.getElementById("book_name").value = e.target.dataset.title;
      document.getElementById("author_name").value = e.target.dataset.author;
    });
  });
  msg.innerText = "";
}

async function issueBook() {
  const selected = document.querySelector('input[name="book"]:checked');
  const userId = document.getElementById("user_id").value.trim();
  const issueDate = document.getElementById("issue_date").value;
  const returnDate = document.getElementById("return_date").value;
  const remarks = document.getElementById("remarks").value.trim();
  const msg = document.getElementById("msg");

  if (!selected || !userId || !issueDate || !returnDate) {
    msg.innerText = "Please complete all mandatory fields and select a book.";
    return;
  }

  const issue = new Date(issueDate);
  const ret = new Date(returnDate);
  const max = new Date(issueDate);
  max.setDate(max.getDate() + 15);

  if (issueDate < todayISO()) {
    msg.innerText = "Issue date cannot be lesser than today.";
    return;
  }
  if (ret > max || ret < issue) {
    msg.innerText = "Return date must be between issue date and 15 days ahead.";
    return;
  }

  const res = await fetch(`${API}/transactions/issue-book`, {
    method: "POST",
    headers: authHeaders(),
    body: JSON.stringify({
      user_id: Number(userId),
      book_id: Number(selected.value),
      issue_date: issueDate,
      return_date: returnDate,
      remarks: remarks || null,
    }),
  });
  const data = await res.json();
  msg.innerText = res.ok
    ? `Book issued successfully. Transaction ID: ${data.transaction_id}`
    : data.detail || "Issue failed";
}

async function returnBook() {
  const transactionId = document.getElementById("transaction_id").value.trim();
  const serialNo = document.getElementById("serial_no").value.trim();
  const returnDate = document.getElementById("actual_return_date").value;
  const msg = document.getElementById("msg");

  if (!transactionId || !serialNo || !returnDate) {
    msg.innerText = "Transaction ID, serial number and return date are mandatory.";
    return;
  }

  const res = await fetch(`${API}/transactions/return-book`, {
    method: "POST",
    headers: authHeaders(),
    body: JSON.stringify({
      transaction_id: Number(transactionId),
      serial_no: serialNo,
      return_date: returnDate,
    }),
  });
  const data = await res.json();
  if (!res.ok) {
    msg.innerText = data.detail || "Return validation failed";
    return;
  }
  localStorage.setItem("return_txn_id", String(data.transaction_id));
  localStorage.setItem("return_fine", String(data.fine));
  msg.innerText = `Validated. Proceed to Pay Fine page. Fine: ${data.fine}`;
}

async function payFine() {
  const txnId = localStorage.getItem("return_txn_id");
  const finePaid = document.getElementById("fine_paid").checked;
  const remarks = document.getElementById("fine_remarks").value.trim();
  const msg = document.getElementById("msg");

  if (!txnId) {
    msg.innerText = "No return transaction found. Use Return Book first.";
    return;
  }

  const res = await fetch(`${API}/transactions/pay-fine`, {
    method: "POST",
    headers: authHeaders(),
    body: JSON.stringify({
      transaction_id: Number(txnId),
      fine_paid: finePaid,
      remarks: remarks || null,
    }),
  });
  const data = await res.json();
  msg.innerText = res.ok ? data.message : data.detail || "Payment failed";
}

async function loadActiveIssues() {
  const msg = document.getElementById("msg");
  if (msg) msg.innerText = "";
  const res = await fetch(`${API}/transactions/active-issues`, { headers: authHeaders() });
  const data = await res.json();
  const list = document.getElementById("activeList");
  list.innerHTML = "";
  if (!res.ok) {
    if (msg) msg.innerText = data.detail || "Unable to load active issues";
    return;
  }
  if (!data.length) {
    if (msg) msg.innerText = "No active issues found.";
    return;
  }
  data.forEach((t) => {
    list.innerHTML += `<li class="list-group-item">Txn ${t.transaction_id} | User ${t.user_id} | Book ${t.book_id} | Due ${t.due_date}</li>`;
  });
}

async function loadOverdueReturns() {
  const msg = document.getElementById("msg");
  if (msg) msg.innerText = "";
  const res = await fetch(`${API}/transactions/overdue-returns`, { headers: authHeaders() });
  const data = await res.json();
  const table = document.getElementById("overdueTable");
  table.innerHTML = `
    <tr>
      <th>Transaction</th><th>User</th><th>Book</th><th>Due Date</th><th>Days Late</th><th>Fine</th>
    </tr>`;
  if (!res.ok) {
    if (msg) msg.innerText = data.detail || "Unable to load overdue returns";
    return;
  }
  if (!data.length) {
    if (msg) msg.innerText = "No overdue returns found.";
    return;
  }
  data.forEach((t) => {
    table.innerHTML += `<tr><td>${t.transaction_id}</td><td>${t.user_id}</td><td>${t.book_id}</td><td>${t.due_date}</td><td>${t.days_late}</td><td>${t.fine}</td></tr>`;
  });
}

window.addEventListener("DOMContentLoaded", () => {
  setDashboardLink();

  const issueDate = document.getElementById("issue_date");
  const returnDate = document.getElementById("return_date");
  if (issueDate && returnDate) {
    issueDate.value = todayISO();
    returnDate.value = addDaysISO(15);
    issueDate.addEventListener("change", () => {
      returnDate.value = addDaysISO(15);
    });
  }

  const actualReturnDate = document.getElementById("actual_return_date");
  if (actualReturnDate) {
    actualReturnDate.value = todayISO();
  }
});
